### ISSUE TITLE
Query: Use parameterization when Contains() only has a small number of elements

### ISSUE THREAD

**Issue by rmja** [dotnet/efcore#6809]
When filtering using `Contains()` on a collection with only a single item, the generated SQL should treat it as parameterized SQL to minimize the number of execution plans.

For example:

```
var ids = new[] {1,2};
await Context.Cars.Where(x => ids.Contains(x.Id)).ToListAsync();
```

Correctly produces SQL with `WHERE Id IN (1,2)`. However, if the `ids` array only had one item, say `ids = new[] {1}`, it would be nice if the generated SQL did not use `WHERE IN (1)`, but instead equality on the Id, using a parameterized version. Something like `WHERE Id = @p0`.

EF Core version: 1.0.0

**Comment by popcatalin81** [comment]
@rmja for Sql Server 2005 + it does not matter, Queries where the discriminator is the clustered key are auto parametrized to use clustered index seek plan.  If you analyze the execution plans for your examples you will see the same execution plan for all cases.

**Comment by rowanmiller** [comment]
@rmja did you have a specific scenario where this would provide a meaningful performance improvement? We're willing to consider it if there is a tangible benefit.

**Comment by divega** [comment]
As I think @ajcvickers mentioned today in triage we could consider having a threshold (likely a small number greater than 1) under which we could create parameterized queries for the translation of `Enumerable.Contains()`.

Even assuming @popcatalin81's [comment] (https://github.com/aspnet/EntityFramework/issues/6809#issuecomment-254776796) is correct this should still help when the query is not auto-parameterized (e.g. if is not on a clustered index), with other providers besides SQL Server and hopefully to avoid regenerating the SQL inside our own stack (see #4605 for some context). 

This sounds like a potentially lower impact approach that would still yield a benefit in many common scenarios compared to other options we have considered in the past such as using TVPs.

**Comment by rmja** [comment]
Yes, having a per-context configurable threshold sounds like a good approach.

**Comment by divega** [comment]
@smitpatel do you think this can be closed now as a duplicate of #12777 that you filed? It seems the latter covers more than the case in which the number of items is relatively small.

**Comment by smitpatel** [comment]
Duplicate of #12777
